docs/05-trees.md

```diff
@@ -0,0 +1,350 @@
+
+```moonbit no-check
+enum AVLTree {
+  Empty
+  // current value, left subtree, right subtree, height
+  Node(Int, AVLTree, AVLTree, Int) 
+}
+fn create(value: Int, left: AVLTree, right: AVLTree) -> AVLTree
+fn height(tree: AVLTree) -> Int
+```
+
+The creation function creates a new AVL tree without explicitly maintaining its height. Since the insertion and deletion operations of AVL trees are similar to standard binary search trees, we won't go into detail here.
+
+![](/pics/rotation.drawio.webp)
+
+After inserting or deleting an element, we traverse back from the modified location until we find the first unbalanced position, which we call $z$. Then, we use $y$ to represent the higher subtree of $z$, and $x$ is the higher subtree of $y$. Next, we discuss rebalancing. In the first case, $x$ is between $y$ and $z$. In this case, we can move $x$ above its parent and grandparent. We can see that the depths of $x$'s two subtrees decrease by $1$, thereby reducing the height of the entire tree. Although the depth of $T_4$ increases by $1$, it was originally lower than the left subtree, so it remains balanced. The other case is when $x$ is on $y$ and $z$'s same side. In this case, we can reduce the height of the tree by making $y$ the root node. The purpose is still to reduce the depths of the two deepest subtrees to decrease the height of the left subtree.
+
+```moonbit no-check
+fn balance(left: AVLTree, z: Int, right: AVLTree) -> AVLTree {
+  if height(left) > height(right) + 1 {
+    match left {
+      Node(y, left_l, left_r, _) => 
+        if height(left_l) >= height(left_r) { 
+          create(left_l, y, create(lr, z, right)) // x is on y and z's same side 
+        } else { match left_r {
+          Node(x, left_right_l, left_right_r, _) =>  // x is between y and z
+            create(create(left_l, y, left_right_l), x, create(left_right_r, z, right))
+        } }
+    }
+  } else { ... }
+}
+```
+
+Here is a snippet of code for a balanced tree. You can easily complete the code once you understand what we just discussed. We first determine if a tree is unbalanced, by checking if the height difference between the two subtrees exceeds a specific value and which side is higher. After determining this, we perform a rebalancing operation. At this point, the root node we pass in is $z$, and the higher side after pattern matching is $y$. Then, based on the comparison of the heights of $y$'s two subtrees, we further determine whether $x$ is $y$ and $z$'s same side or between $y$ and $z$, as shown in line 6. Afterwards, we recombine based on the scenarios discussed earlier, as shown in lines 6 and 9. Taking insertion of an element as an example: 
+
+```moonbit no-check
+fn add(tree: AVLTree, value: Int) -> AVLTree {
+  match tree {
+    Node(v, left, right, _) as t => {
```

Comment:
Maybe we can introduce about label in enum in #22 and rewrite this part as something like
```
match tree {
  Node(v, ..) as t => {
    if value < v { balance(add(t.left, value), v, t.right) } else { ... }
  }
}
```
where
```
enum AVLTree {
  Empty
  Node(Int, ~left: AVLTree, ~right: AVLTree, ~height: Int)
}
```